### DISCUSSION TITLE
Defining new functions in ParameterValues

### DISCUSSION THREAD

**Discussion by ahsadeghi** [pybamm-team/PyBaMM#3103] [Q&A] [answered]
Hello,
I am trying to define a new function for calculating (positive) current collector conductivity as

```
def positive_current_collector_conductivity(T):   
    #Strum(2019)
    sigma_cc = 3.78e7 / (1 + 4.20e-3 * (T - 293.15))
    return sigma_cc
```

in `ORegan2022.py` parameters, and I call the function in the end of the file, return { , as 

`"Positive current collector conductivity [S.m-1]": positive_current_collector_conductivity,
`

but I get  the following error

`TypeError: Cannot process parameter '<function positive_current_collector_conductivity at 0x000001E376B9F310>'
`

I wonder if  you  would please help to  find the reason. Thank you.

Regards,
Hosein

**Comment by ahsadeghi** [comment]
Dear @brosaplanella 
Do you have any comments regarding this issue?

**Comment by valentinsulzer** [comment, marked as answer]
This is because the current collector conductivity is a `Parameter`, not a `FunctionParameter` 
It needs to be change from [this](https://github.com/pybamm-team/PyBaMM/blob/6b642a6c3c9d5d90c15f9a807f8bef5998efda6b/pybamm/parameters/lithium_ion_parameters.py#L282) to [this](https://github.com/pybamm-team/PyBaMM/blob/6b642a6c3c9d5d90c15f9a807f8bef5998efda6b/pybamm/parameters/lithium_ion_parameters.py#L345)

This should be a super easy fix if you want to try it

**Comment by ganmad** [comment]
Hi @tinosulzer,

As per your suggestion, I tried to implement temperature dependent EC diffusivity, by replacing   `self.D_ec = pybamm.Parameter(f"{pref}EC diffusivity [m2.s-1]")` 
with 
```
def D_eci(self,T):
                pref = self.phase_prefactor
                return pybamm.FunctionParameter(f"{pref}EC diffusivity [m2.s-1]", {"Temperature [K]": T})

self.D_ec = D_eci
```
I tried different ways, defining the function outside/inside of the `_set_parameters`, in lithium_ion_parameters.py, but doesn't work. I defined a temperature dependence function locally as well to update parameters. Am I missing something here? You inputs will be much appreciated. Thanks.

**Reply by brosaplanella** [reply]
You need to define it as a function of the class `ParticleLithiumIonParameters` so the `self.D_ec = D_eci` is not needed anymore (see, for example, how it is done with particle diffusivity: https://github.com/pybamm-team/PyBaMM/blob/f81de9416904b4bd36eb75434f37660c84a79bdc/pybamm/parameters/lithium_ion_parameters.py#L537-L560).

You will also have to pass the temperature as an argument everywhere where this parameter is called, but you can do a search in the code base to help you with this.

Note that #3198 does some changes in this part of the code, so you might want to wait until that one is merged (which should happen this week).

**Comment by ganmad** [comment]
Hi @brosaplanella,

As suggested, I tried implementing a method for D_ec(T) as below: 

```
def D_ec(self, T):
        """Dimensional EC diffusivity as f(T)""" 
        #main = self.main_param
        #domain, Domain = self.domain_Domain
        #phase_name = self.phase_name
        pref = self.phase_prefactor
        if self.domain == "negative":
            inputs = {"Temperature [K]": T}
            #Domain = self.domain.capitalize()
            return pybamm.FunctionParameter(
                f"{pref}EC diffusivity [m2.s-1]", inputs
           )

```
Also, I commented the D_ec line at [l454](https://github.com/pybamm-team/PyBaMM/blob/f81de9416904b4bd36eb75434f37660c84a79bdc/pybamm/parameters/lithium_ion_parameters.py#L454C13-L454C13)

But, when I load the model with the following options, 
`model = pybamm.lithium_ion.SPM({"cell geometry": "arbitrary","thermal": "lumped","SEI":"ec reaction limited"})`

I get this NotImplementedError:

![image](https://github.com/pybamm-team/PyBaMM/assets/50149439/31a77af6-1ae9-4cb8-a8a7-177a68b1910c)


Do I have to do anything after defining the method` D_ec(self,T)`? 

Thank you.

**Reply by brosaplanella** [reply]
Where did you place this new function? It should be in `lithium_ion_parameters.py`. But you also will have to replace all occurrences of `D_EC` with D_EC(T), where T needs to be the right temperature for the domain. If are planning on getting this merged to PyBaMM, you can open a pull request so I can see all the code and give more specific comments.

**Reply by ganmad** [reply]
Hi @brosaplanella ,

I managed to implement this by replacing D_ec with D_ec(T) at all instances. So it is working fine. Thanks for your suggestion. Regards,

**Comment by Brushnie33** [comment]
@brosaplanella Hello, I managed to change D_ec to D_ec(T) and get it to work. However, if I follow a similar approach to define inner_sei_proportion to be SOC dependent, I got errors. In lithium_ion_parameters.py, I defined the following equation inside class ParticleLithiumIonParameters(BaseParameters):

    def cracking_tuning_factor(self, sto):
        domain, Domain = self.domain_Domain
        inputs = {f"{Domain} particle stoichiometry": sto},
        return pybamm.FunctionParameter(
            f"{self.phase_prefactor}{Domain} electrode cracking reaction tuning factor", inputs
        )

And in the sei_growth.py codes, I added the following codes:
       sto = variables[f"{Domain} particle concentration"] # variables[f"{Domain} {self.phase_name}particle concentration"] 
        inner_sei_proportion= phase_param.cracking_tuning_factor(sto)

I believe the errors are associated with sto, could you please suggest how to modify sto in the above codes?
